app/controllers/admin/badges_controller.rb

```diff
@@ -62,20 +71,63 @@ def mass_award
       mode = Email.is_valid?(CSV.parse_line(csv.first).first) ? 'email' : 'username'
       csv.rewind
 
+      if grant_existing_holders
```

Comment:
A bit of a distraction/suggestion, I feel like this controller action is getting too big. Perhaps it is time to split the logic here into a service object?

Reply:
Now that I've reduced the new logic that I need to add in the controller action to roughly half what it was before, do you still feel we should still reduce the controller action and extract things into a service object?